File: core/src/com/alta_v2/game/ScreenManager.java
package com.alta_v2.game;

import com.alta_v2.game.screen.GameScreen;
import com.alta_v2.game.screen.GameScreenFactory;
import com.alta_v2.mediatorModule.screen.ScreenContext;
import com.badlogic.gdx.Screen;
import com.google.inject.Inject;
import lombok.Setter;
import lombok.extern.log4j.Log4j2;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Provides the manager to handle screens.
 */
@Log4j2
public class ScreenManager {

    @Setter
    private AltaV2 game;

    private final GameScreenFactory screenFactory;
    private final AtomicBoolean screenChangeIsLocked;

    /**
     * Initialize new instance of {@link ScreenManager}.
     *
     * @param screenFactory - the {@link GameScreenFactory} instance.
     */
    @Inject
    public ScreenManager(GameScreenFactory screenFactory) {
        this.screenFactory = screenFactory;
        this.screenChangeIsLocked = new AtomicBoolean(false);
    }

    /**
     * Changes the screen.
     *
     * @param screenContext - the {@link ScreenContext} instance.
     */
    public void changeScreen(ScreenContext screenContext) {
        if (this.game == null) {
            throw new NullPointerException("AltaV2 must not be null");
        }

        if (this.screenChangeIsLocked.get()) {
            log.warn("Screen changing is not available because another change operation in progress.");
            return;
        }

        try {
            this.screenChangeIsLocked.set(true);;
            GameScreen oldScreen = this.getScreenAsType(GameScreen.class);
            GameScreen newGameScreen = this.screenFactory.createGameScreen(screenContext);

            if (oldScreen == null) {
                this.game.setScreen(newGameScreen);
                this.screenChangeIsLocked.set(false);
            } else {
                oldScreen.fadeOutScreen(() -> {
                    this.game.setScreen(newGameScreen);
                    oldScreen.dispose();
                    this.screenChangeIsLocked.set(false);
                });
            }
        } catch (Exception e) {
            log.error(e);
            this.screenChangeIsLocked.set(false);
        }
    }

    /**
     * Gets the screen as one of child type.
     *
     * @param resultType - the type that will returns.
     * @return the resultType instance or null;
     */
    private  <T extends Screen> T getScreenAsType(Class<T> resultType) {
        if (this.game == null) {
            throw new NullPointerException("AltaV2 must not be null");
        }

        Screen screen = this.game.getScreen();
        if (!this.isInstanceOf(resultType, screen)) {
            return null;
        }

        return resultType.cast(screen);
    }

    private <T> boolean isInstanceOf(Class<T> clazz, Screen screen) {
        return clazz.isInstance(screen);
    }
}

File: core/src/com/alta_v2/game/ScreenManager.kt
package com.alta_v2.game

import com.alta_v2.exception.ChangeScreenException
import com.alta_v2.game.screen.GameScreen
import com.alta_v2.game.screen.GameScreenFactory
import com.alta_v2.game.utils.ChangeScreenResult
import com.alta_v2.mediator.screen.context.ScreenContext
import com.google.inject.Inject
import mu.KotlinLogging
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Provides the manager to handle screens.
 */
class ScreenManager @Inject constructor(private val screenFactory: GameScreenFactory) {

    private val log = KotlinLogging.logger {  }

    private var game: AltaV2? = null
    private val screenChangeIsLocked: AtomicBoolean = AtomicBoolean(false)

    fun setGame(game: AltaV2) {
        this.game = game
    }

    /**
     * Changes the screen.
     *
     * @param screenContext - the [ScreenContext] instance.
     */
    fun changeScreen(screenContext: ScreenContext?): ChangeScreenResult {
        if (game == null) {
            throw ChangeScreenException("AltaV2 container is null but must not")
        }
        if (screenChangeIsLocked.get()) {
            log.warn("Screen changing is not available because another change operation in progress.")
            throw ChangeScreenException("Screen changing is not available because another change operation in progress.")
        }
        val changeScreenResult = ChangeScreenResult()
        try {
            screenChangeIsLocked.set(true)
            val newGameScreen = screenFactory.createGameScreen(screenContext!!)

            if (game?.screen == null) {
                game?.screen = newGameScreen
                screenChangeIsLocked.set(false);
            } else {
                val oldScreen = game?.screen as GameScreen
                oldScreen.fadeOutScreen(Runnable {
                    game?.screen = newGameScreen
                    oldScreen.dispose()
                    screenChangeIsLocked.set(false)
                    changeScreenResult.complete()
                })
            }
        } catch (e: Exception) {
            screenChangeIsLocked.set(false)
            throw ChangeScreenException(e)
        }
        return changeScreenResult
    }
}